dll/win32/uxtheme/nonclient.c

```diff
@@ -1075,6 +1078,125 @@ ThemeWndProc(HWND hWnd, UINT Msg, WPARAM wParam, LPARAM lParam, WNDPROC DefWndPr
+        SetScrollInfo(pcontext->hWnd, SB_VERT, &dummyScrollInfo, TRUE);
+    }
+
+    RECT rcWindowPrev = { pcontext->wi.rcWindow.left, pcontext->wi.rcWindow.top, pcontext->wi.rcWindow.right, pcontext->wi.rcWindow.bottom };
+    UNREFERENCED_PARAMETER(rcWindowPrev);
+    RECT rcClientPrev = { pcontext->wi.rcClient.left, pcontext->wi.rcClient.top, pcontext->wi.rcClient.right, pcontext->wi.rcClient.bottom };
+    UNREFERENCED_PARAMETER(rcClientPrev);
+    SetWindowPos(pcontext->hWnd, NULL, left, top, right - left, bottom - top, SWP_NOZORDER | SWP_NOACTIVATE | SWP_DRAWFRAME | SWP_NOCOPYBITS);
+    RECT rcWindowNew = { left, top, right, bottom };
+    pcontext->wi.rcWindow = rcWindowNew;
+
+    SetViewportOrgEx(hDC, rcWindowNew.left, rcWindowNew.top, NULL);
+
+    INT offsetX = -rcWindowNew.left;
+    INT offsetY = -rcWindowNew.top;
+    OffsetRect(&rcWindowNew, offsetX, offsetY);
+    ThemeCalculateCaptionButtonsPosEx(pcontext->wi, pcontext->hWnd, pcontext->theme, pcontext->CaptionHeight - pcontext->wi.cyWindowBorders);
+
+
+    INT leftBorderInset = pcontext->wi.cxWindowBorders;
+    INT titleBarInset = pcontext->CaptionHeight; // + pcontext->wi.cyWindowBorders;
+    INT rightBorderInset = pcontext->wi.cxWindowBorders;
+    INT bottomBorderInset = pcontext->wi.cyWindowBorders;
+    UNREFERENCED_PARAMETER(leftBorderInset);
+    UNREFERENCED_PARAMETER(titleBarInset);
+    UNREFERENCED_PARAMETER(rightBorderInset);
+    UNREFERENCED_PARAMETER(bottomBorderInset);
+
+    RECT rcClientNew;
+    if (GetWindowRect(pcontext->hWnd, &rcClientNew))
+    {
+        rcClientNew.left += leftBorderInset;
+        rcClientNew.top += titleBarInset;
+        rcClientNew.right -= rightBorderInset;
+        rcClientNew.bottom -= bottomBorderInset;
+    }
+    pcontext->wi.rcClient = rcClientNew;
+
+    pcontext->wi.dwStyle &= ~(WS_HSCROLL | WS_VSCROLL);
+    ThemePaintWindow(pcontext, &rcWindowNew, FALSE);
+    pcontext->wi.dwStyle = dwStyle;
+
+    if (hasVScrollBar && IsScrollBarVisible(pcontext->hWnd, OBJID_VSCROLL))
+    {
+        SCROLLBARINFO sbi;
+        sbi.cbSize = sizeof(sbi);
+        GetScrollBarInfo(pcontext->hWnd, SCROLL_getObjectId(SB_VERT), &sbi);
+        INT scWidth = sbi.rcScrollBar.right - sbi.rcScrollBar.left;
+
+        sbi.rcScrollBar.right = rcClientNew.right;
+        rcClientNew.right -= scWidth;
+        sbi.rcScrollBar.left = rcClientNew.right;
+
+        sbi.rcScrollBar.top = rcClientNew.top;
+        sbi.rcScrollBar.bottom = rcClientNew.bottom;
+
+        ThemeDrawScrollBarEx(pcontext, SB_VERT, &sbi, NULL);
+    }
+    pcontext->wi.rcClient = rcClientNew;
+
+    SelectObject(hDC, GetThemeSysColorBrush(pcontext->theme, COLOR_WINDOW));
+    HPEN pen = CreatePen(PS_NULL, 0, RGB(0,0,0));
+    SelectObject(hDC, pen);
+
+    OffsetRect(&rcClientNew, -pcontext->wi.rcWindow.left, -pcontext->wi.rcWindow.top);
+    if (!(dwExStyle & WS_EX_DLGMODALFRAME))
+    {
+        rcClientNew.right++;
+        rcClientNew.bottom++;
```

Comment:
Should this come from the theme?

Reply:
I...don't think so? Truth be told, I'm not 100% sure - there is *definitely* something wrong with how our window frames work compared to Windows's, which made the already-difficult task of sifting through `win32ss/user/**` for insight even harder, since now I don't even know if I'm just imitating incorrect behaviour. I've been trying to make sense of all that on and off for a while now, in the hopes of fixing our window frames...no luck so far though. I don't recall 100% for sure whether this was one of the things I gleamed from there...but if so, I'd have followed suit for now, since I've yet to identify a more correct method.

Reply:
Is our uxtheme dll able to run in Windows itself? (and same question for some of our themes?) If so, what about testing that on Windows to be sure that this code works there? If then put back into ReactOS and there exists visual differences, this would point to a problem in the code in win32ss/user/* .
(I also wonder whether Windows' own uxtheme dll could run in ReactOS. Maybe so if we consider the Win2k3 one?)

Reply:
@HBelusca No, unfortunately, our `uxtheme.dll` isn't functional under Windows at this point in time.

However, I *was* able to run our `desk.cpl` on Windows Server 2003 quite easily, and that's the part that actually calls `DrawNCPreview`. Though this was useful for my purposes, it doesn't exactly help with what you're getting at. That said, I think the problems with our window frames are out of scope of this PR - after all, those issues aren't specific to the visual style preview, they affect "real" windows as well.